### PR TITLE
Added state machine to MatchState for restricting Simulations.

### PR DESCRIPTION
**Pull request by LadyEbony**
### Motivation and Context
- [X] Why is this change required? What problem does it solve?
Timer was playing during countdown and game results could theoretically change after the EndEvent but before stage switch. Furthermore, MatchManager was getting too many variables to represent the match's state. This fix addresses it by having "basically global" MatchStateID that represents the current state which can accessed by any script.
- [ ] If it fixes an open issue, include the text `Closes #1` (where 1 would be the issue number) to your commit message.

### Types of changes
What types of changes does your code introduce? Check all the boxes that apply:
- [X] Bug fix (non-breaking change which fixes an issue)
Timer doesn't decrease during countdown.
- [X] Breaking change (fix or feature that would cause existing functionality to change)
MatchState now contains a StateID which represents the Match's, well, current state. Accessible.
What simulations called in MatchSimulation can be altered based on the MatchStateID.

### Description
- [X] Describe your changes in detail.
The commits and change list kinda explain it all. I'm not a too big of my implementation. Seems a little unclean.

### CODE REVIEW

File: Assets/Code/src/Runtime/Match/Components/MatchManager.cs [hunk 1]
```diff
@@ -51,16 +48,16 @@ public class MatchManager : MonoBehaviour {
...
         MatchState = MatchController.CurrentState
     });
-    IsControllerRunning = true;
-
-    Debug.Log("Running match...");
-    MatchTask = new TaskCompletionSource<MatchResult>();
+		
+	 Debug.Log("Running match...");
+	 MatchController.CurrentState.StateID = MatchStateID.InGame;
+	 MatchTask = new TaskCompletionSource<MatchResult>();
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/Components/MatchManager.cs]: The style for this project uses spaces, please change the tabs to spaces instead.

File: Assets/Code/src/Runtime/Match/Components/MatchPauseController.cs [hunk 2]
```diff
@@ -25,9 +25,9 @@ public class MatchPauseController : MonoBehaviour {
   /// </summary>
   void Update() {
     if (MatchManager == null || !MatchManager.IsLocal) return;
-    if (MatchManager.IsPaused) {
+    if (MatchManager.MatchController.CurrentState.StateID == MatchStateID.Pause) {
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/Components/MatchPauseController.cs]: MatchController.CurrentState.StateID is being called a lot, might be better if there was property that made it easier to access.

File: Assets/Code/src/Runtime/Match/MatchSimulation.cs [hunk 3]
```diff
@@ -39,6 +44,14 @@ public class MatchSimulation : IMatchSimulation {
     }
   }
 
-}
+  void SimulateRestrict(ref MatchState state, MatchInputContext input, params Type[] restrictions) {
+	 foreach (var SimComponent in SimulationComponents) {
+		if (restrictions.Contains(SimComponent.GetType()))
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/MatchSimulation.cs]: Calling GetType on and comparing against types every simulation (60FPS) can get very expensive and may allocate memory. 

Suggested fix: Create separate IMatchSimulation[] arrays for each MatchStateID, put them into a Dictionary<MatchStateID, IMatchSimulation[]> and retrieve the simulation set via state.StateID, then simulate. Uses marginally more permanent memory for less GC allocation and faster execution.

`simulations[state.StateID].Simulate(state, inputContext)`

File: Assets/Code/src/Runtime/Match/Components/MatchManager.cs [hunk 4]
```diff
@@ -72,13 +69,16 @@ public class MatchManager : MonoBehaviour {
...
+	 MatchController.CurrentState.StateID = MatchStateID.End;
+	 return result;
   }
 
   public void SetPaused(bool paused) {
     if (!IsLocal) return;
-    bool changed = paused != IsPaused;
-    IsPaused = paused;
+
+	 var pauseID = paused ? MatchStateID.Pause : MatchStateID.InGame;
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/Components/MatchManager.cs]: Needs an additional check: you can only pause if StateID is InGame, if it's not Paused or InGame, you shouldn't be doing either.

File: Assets/Code/src/Runtime/Match/Components/MatchManager.cs [hunk 5]
```diff
@@ -31,9 +28,9 @@ public class MatchManager : MonoBehaviour {
   /// This function is called every fixed framerate frame, if the MonoBehaviour is enabled.
   /// </summary>
   void FixedUpdate() {
-    if (!IsPaused) {
+    if (MatchController.CurrentState.StateID != MatchStateID.Pause)
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/Components/MatchManager.cs]: Style: Keep the surrounding curly braces.

File: Assets/Code/src/Runtime/Match/MatchState.cs [hunk 6]
```diff
@@ -7,6 +7,8 @@
 
 namespace HouraiTeahouse.FantasyCrescendo.Matches {
 
+public enum MatchStateID { Intro, InGame, Pause, End }
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/MatchState.cs]: Document this and each of the enum values.

Also not sure about the naming of MatchStateID, doesn't seem right, but MatchState is already being used. MatchProgressionStage?

File: Assets/Code/src/Runtime/Match/MatchState.cs [hunk 7]
```diff
@@ -17,6 +19,7 @@ public class MatchState : INetworkSerializable {
 
   PlayerState[] playerStates;
   public int PlayerCount { get; private set; }
+  public MatchStateID StateID { get; set; } = MatchStateID.Intro;
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/MatchState.cs]: If it doesn't need to be hidden, for these small data objects like this, it's better to just use a public field.

This also should be serialized/deserialized in Serialize/Deserialize below. If you do not know how to do it, leave a TODO for me and I'll get around to doing it.

File: Assets/Code/src/Runtime/Core/GameController.cs [hunk 8]
```diff
@@ -19,19 +19,18 @@ public class MatchController : IMatchController {
...
+		var input = InputSource.SampleInput();
+		Assert.IsTrue(input.IsValid);
+		inputContext.Update(input);
+	 }
+    
+	 var state = CurrentState;
+	 Simulation.Simulate(ref state, inputContext);
+	 CurrentState = state;
+
+	 Timestep++;
```
Comment (james7132) [inline Assets/Code/src/Runtime/Core/GameController.cs]: Style: four space indent -> two space indent.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Comment by LadyEbony** [comment]
Address requested changes.
Check commit for more details. (Forget to mention that I added serialization for MatchProgressionState)

File: Assets/Code/src/Runtime/Core/GameController.cs [hunk 9]
```diff
@@ -10,23 +10,27 @@ public class MatchController : IMatchController {
 
   public virtual uint Timestep { get; set; }
   public virtual MatchState CurrentState { get; set; }
+  public virtual MatchProgressionState CurrentProgressionID {
```
Comment (james7132) [inline Assets/Code/src/Runtime/Core/GameController.cs]: I should have clarified: put this on MatchManager, not MatchController.

File: Assets/Code/src/Runtime/Match/MatchSimulation.cs [hunk 10]
```diff
@@ -12,25 +11,43 @@ public interface IMatchSimulation : IInitializable<MatchConfig>, ISimulation<Mat
 
 public class MatchSimulation : IMatchSimulation {
 
+  // Contains all simulations
   readonly IMatchSimulation[] SimulationComponents;
+  // Contains simulations pertaining to the MatchProgressionState
+  Dictionary<MatchProgressionState, IMatchSimulation[]> SimulationDictionary;
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/MatchSimulation.cs]: No need to add Dictionary to the name, the declaration for it already makes it clear. Simulations is declarative enough.

File: Assets/Code/src/Runtime/Match/MatchSimulation.cs [hunk 11]
```diff
@@ -12,25 +11,43 @@ public interface IMatchSimulation : IInitializable<MatchConfig>, ISimulation<Mat
...
 
   public MatchSimulation(IEnumerable<IMatchSimulation> simulationComponents) {
     SimulationComponents = simulationComponents.ToArray();
-  }
+
+    SimulationDictionary = new Dictionary<MatchProgressionState, IMatchSimulation[]>();
+    SimulationDictionary.Add(
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/MatchSimulation.cs]: More readable if you use a collection initializer: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer

File: Assets/Code/src/Runtime/Match/Components/MatchPauseController.cs [hunk 12]
```diff
@@ -6,7 +6,7 @@ namespace HouraiTeahouse.FantasyCrescendo {
 
 public class MatchPauseController : MonoBehaviour {
 
-  public MatchManager MatchManager;
+  public MatchManager MatchManager { get { return MatchManager.Instance; } }
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/Components/MatchPauseController.cs:9]: Style: Please use expression bodied statement: `MatchManager =>MatchManager.Instance` for get-only one-liner properties.

File: Assets/Code/src/Runtime/Match/MatchSimulation.cs [hunk 13]
```diff
@@ -39,6 +56,16 @@ public class MatchSimulation : IMatchSimulation {
     }
   }
 
+  IMatchSimulation[] CreateSimulations(IMatchSimulation[] matchSimulations, params Type[] restrictions) {
+    var Simulation = new List<IMatchSimulation>();
+    foreach (var simComponent in matchSimulations) {
```
Comment (james7132) [inline Assets/Code/src/Runtime/Match/MatchSimulation.cs]: This is reducible via Linq to `matchSimulations.Where(sim => restrictions.Contains(sim)).ToArray()`

**Review by james7132 — CHANGES_REQUESTED** [review]
Clarified one of my previous comments. Mostly style and naming changes otherwise.

**Comment by LadyEbony** [comment]
Addressed code style fixes.

**Review by james7132 — APPROVED** [review]

**Pull request merged**